### PR TITLE
Use pyface < 7

### PR DESCRIPTION
**Pull request by cbrnr**
Fixes (works around) #8040 hopefully.

### CODE REVIEW

**Comment by cbrnr** [comment]
This seems to work, but it would be better if we used the most recent `pyface` (7.0.1). The problem is that for some reason, `traits` is pinned to 5.2.0, but `pyface` 7.0.1 requires the latest `traits` 6.1.0 (which is available in the default channel).

File: environment.yml [hunk 1]
```diff
@@ -18,13 +18,15 @@ dependencies:
...
 - numexpr
-- traits>=4.6.0
-- pyface>=6
-- traitsui>=6
-- imageio>=2.6.1
+- traits
+- pyface>=6,<7
+- traitsui
+- imageio
+- numpydoc
```
Comment (larsoner) [inline environment.yml:25]: We can just get rid of this actually, but that can be another PR. This one seems to be headed toward green (other than our recent pip-pre 3.8 segfaults that are unrelated)

Comment (cbrnr) [reply environment.yml:25]: What do you mean? Remove these four packages completely from the env?

Comment (larsoner) [reply environment.yml:25]: No just `numpydoc`

**Comment by larsoner** [comment]
Do we know which package actually requires `pyface<7`?

File: environment.yml (hunk 1 above)
Comment (cbrnr) [reply environment.yml:25]: OK, I'll remove `numpydoc`. What about the other packages I moved from `pip` to `conda`? OK or revert?

**Comment by cbrnr** [comment]
> Do we know which package actually requires `pyface<7`?

No, I tried to let `conda` solve the conflict, but after working for 15 minutes it spat out a very long list of conflicts that didn't really help me (because every package seemed to conflict with several other packages).

**Comment by larsoner** [comment]
Which OS? I tried on macOS and Linux and in both cases `conda update conda` followed by `conda create -n test -f environment.yml` from `master` worked, in both cases installing pyface 7.0.1. So maybe this was just a transient issue?

**Comment by larsoner** [comment]
To check, maybe remove the `numpydoc` and the `<7` pin from `pyface` and see if it succeeds here?

File: environment.yml [hunk 2]
```diff
@@ -18,13 +18,15 @@ dependencies:
...
+- tqdm
+- spyder-kernels
```
Comment (larsoner) [inline environment.yml:27]: Not sure if we can do this yet, have you checked with #8009 ?

Comment (larsoner) [reply environment.yml:27]: Looks like we probably can since #8009 mentions 1.9.2 and that's what's in conda now

Comment (larsoner) [reply environment.yml:27]: ... but I guess it should be:
```suggestion
- spyder-kernels>=1.9.2
```

Comment (cbrnr) [reply environment.yml:26]: Shouldn't `conda` always install the latest available version unless there is a conflict with other (installed) packages? Let's see if CIs come back happy - if we don't test for the behavior in #8009 I will manually test it.

Comment (larsoner) [reply environment.yml:26]: I think it does when creating from scratch but I'm not sure what it does if you do `conda env update -f environment.yml` or something similar. But maybe we don't need to worry too much about this case

Comment (cbrnr) [reply environment.yml:26]: Yes, I would only support creating the env from scratch, everything else will drive us directly into dependency hell.

**Comment by cbrnr** [comment]
Nope, doesn't work. The error log mentions `qt4 = traitsui.qt4:toolkit` - do we require Qt4 (instead of Qt5) somewhere?

**Comment by larsoner** [comment]
If you change it to `traits>=6` it should hopefully fail earlier. It's in win64:

https://anaconda.org/anaconda/traits

So some package must require it. I'll fire up the Windows VM for local debugging...

**Comment by cbrnr** [comment]
Also, how did `conda create -n test -f environment.yml` (as you've mentioned [here](https://github.com/mne-tools/mne-python/pull/8042#issuecomment-662453071)) work? First, `-f` should be `--file`, but even then do I get an error because the file wants to install in an env called `mne`. That's why I have to use `conda env update --file environment.yml`.

**Comment by larsoner** [comment]
Sorry I meant `conda env create` not `conda create`

**Comment by cbrnr** [comment]
> Sorry I meant `conda env create` not `conda create`

Thanks, this works (pretty inconsistent argument naming I would say). It does give the same results as `conda env update --file environment.yml` though, which is `traits==5.2.0` on my macOS machine.

Why do you think this is related to Windows? The latest version is on conda for all platforms, and we get the issue on a Linux CI (and on my local macOS).

**Comment by larsoner** [comment]
I thought it was installing 6 on macOS and Linux but it wasn't. Changing it to have `- traitsui>=6` and `- traits>=6` fixes things for me on Linux (`pytest mne/gui` didn't work before but does now). No idea why it tries to pull `traits==5.2.0` if you don't tell it to use 6+. For some reason it installs Python 3.6 though...

**Comment by larsoner** [comment]
Adding a `python>=3.8` or even `python>=3.7` does create conflicts. Argh...

**Comment by cbrnr** [comment]
This whole conda thing is a mess. We mix/overwrite with pip because the default channel lags behind recent versions. Should we consider moving to `conda-forge` as @hoechenberger has repeatedly suggested? It looks like such a situation can occur anytime with our env file with the defaults channel.

**Comment by cbrnr** [comment]
(Moving to `conda-forge` also has the advantage that we can use `mamba` instead of `conda`, which is much faster and has more features).

**Comment by larsoner** [comment]
> Should we consider moving to conda-forge as @hoechenberger has repeatedly suggested? It looks like such a situation can occur anytime with our env file with the defaults channel.

I've seen problems with `conda-forge` due to packages not being ready for prime time in other repos that use `conda-forge`, so I wouldn't think of this as something that will solve our problems. It might fix more problems than it creates, but it might also be the reverse.

In any case, this seems to be some `base` problem. Even this causes a conflict:
```
name: mne
channels:
- defaults
dependencies:
- python>=3.8
- traits>=6
- pyface>=7
- traitsui>=6
```
So maybe we should manually pin `pyface` until they sort it out, and raise an upstream issue.

**Comment by cbrnr** [comment]
Agreed - can you raise the upstream issue? You already have a great minimal example. I can pin `pyface` to the old version in the meantime.

**Comment by cbrnr** [comment]
It would still be worth giving `conda-forge` a try, I'd assume that it will solve more problem than it creates (especially the whole `pyqt`/`PyQt5` `conda`/`pip` business).

**Comment by hoechenberger** [comment]
> It would still be worth giving `conda-forge` a try, I'd assume that it will solve more problem than it creates (especially the whole `pyqt`/`PyQt5` `conda`/`pip` business).

Yes this could definitely be worth a shot. The current approach with conda + pip leads to inconsistent environments that behave in the strangest manner if you try to install or update any conda packages. Not a fan.

**Comment by larsoner** [comment]
Thanks @cbrnr

**Comment by larsoner** [comment]
https://github.com/conda/conda/issues/10095

**Comment by drammock** [comment]
by the way, it would have been fine to specify `spyder-kerneles>=1.9.2` as @larsoner suggested.  It just wouldn't actually fix the problem with macOS Catalina.  But no need to go back and do it; it's already WIP in #8020
